crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,738 @@
+        let cargs2 =
+            cargs2.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
+        self.commute_nodes(
+            py,
+            &DAGOpNode {
+                instruction: CircuitInstruction::py_new(
+                    op1,
+                    Some(qargs1.into_any()),
+                    Some(cargs1.into_any()),
+                )?,
+                sort_key: PyString::new_bound(py, "do_not_use").into_any().into(),
+            },
+            &DAGOpNode {
+                instruction: CircuitInstruction::py_new(
+                    op2,
+                    Some(qargs2.into_any()),
+                    Some(cargs2.into_any()),
+                )?,
+                sort_key: PyString::new_bound(py, "do_not_use").into_any().into(),
+            },
+            max_num_qubits,
+        )
+    }
+
+    #[pyo3(signature=())]
+    fn num_cached_entries(&self) -> usize {
+        self.current_cache_entries
+    }
+    #[pyo3(signature=())]
+    fn clear_cached_commutations(&mut self) {
+        self.clear_cache()
+    }
+
+    fn __getstate__(&self, py: Python) -> PyResult<Py<PyDict>> {
+        let out_dict = PyDict::new_bound(py);
+        out_dict.set_item("cache_max_entries", self.cache_max_entries)?;
+        out_dict.set_item("current_cache_entries", self.current_cache_entries)?;
+        out_dict.set_item("_cache_miss", self._cache_miss)?;
+        out_dict.set_item("_cache_hit", self._cache_hit)?;
+        out_dict.set_item("cache", self.cache.clone())?;
+        out_dict.set_item("library", self.library.clone())?;
+        out_dict.set_item("gates", self.gates.clone())?;
+        Ok(out_dict.unbind())
+    }
+
+    fn __setstate__(&mut self, py: Python, state: PyObject) -> PyResult<()> {
+        let dict_state = state.downcast_bound::<PyDict>(py)?;
+        self.cache_max_entries = dict_state
+            .get_item("cache_max_entries")?
+            .unwrap()
+            .extract()?;
+        self.current_cache_entries = dict_state
+            .get_item("current_cache_entries")?
+            .unwrap()
+            .extract()?;
+        self._cache_miss = dict_state.get_item("_cache_miss")?.unwrap().extract()?;
+        self._cache_hit = dict_state.get_item("_cache_hit")?.unwrap().extract()?;
+        self.library = CommutationLibrary {
+            library: dict_state.get_item("library")?.unwrap().extract()?,
+        };
+        self.cache = dict_state.get_item("cache")?.unwrap().extract()?;
+        self.gates = dict_state.get_item("gates")?.unwrap().extract()?;
+        Ok(())
+    }
+}
+
+impl CommutationChecker {
+    #[allow(clippy::too_many_arguments)]
+    fn commute_inner(
+        &mut self,
+        instr1: &CircuitInstruction,
+        qargs1: &Vec<usize>,
+        cargs1: &Vec<usize>,
+        instr2: &CircuitInstruction,
+        qargs2: &Vec<usize>,
+        cargs2: &Vec<usize>,
+        max_num_qubits: u32,
+    ) -> bool {
```

Comment:
Is there a chance that `u32` could be enough to represent `Qarg` indices? The `Qubit` struct is already built on top of a `u32` and `u32 < usize` if you have a 64-bit device, it also comes with the added benefit of not having to cast into a usize in most cases. Also, we could send a slice here to make the method not discriminate against collection types.
```suggestion
    fn commute_inner(
        &mut self,
        instr1: &CircuitInstruction,
        qargs1: &[u32],
        cargs1: &[u32],
        instr2: &CircuitInstruction,
        qargs2: &[u32],
        cargs2: &[u32],
        max_num_qubits: u32,
    ) -> bool {
```